p2p/peer_set_test.go

```diff
@@ -65,3 +70,49 @@ func TestAddRemoveMany(t *testing.T) {
 		}
 	}
 }
+
+func TestPeerSetAddDuplicate(t *testing.T) {
+	t.Parallel()
+	peerSet := NewPeerSet()
+	peer := randPeer()
+	if err := peerSet.Add(peer); err != nil {
+		t.Fatalf("Failed to add new peer: %v", err)
+	}
+
+	var wg sync.WaitGroup
+	for i := 0; i < 10; i++ {
+		// Add them asynchronously to test the
+		// concurrent guarantees of our APIs.
+		wg.Add(1)
+		go func(i int) {
+			defer wg.Done()
+			got, want := peerSet.Add(peer), ErrSwitchDuplicatePeer
+			assert.Equal(t, got, want, "#%d: got=%v want=%v", i, got, want)
+		}(i)
+	}
+	wg.Wait()
+}
+
+func TestPeerSetGet(t *testing.T) {
+	t.Parallel()
+	peerSet := NewPeerSet()
+	peer := randPeer()
+	assert.Nil(t, peerSet.Get(peer.Key), "expecting a nil lookup, before .Add")
+
+	if err := peerSet.Add(peer); err != nil {
+		t.Fatalf("Failed to add new peer: %v", err)
+	}
+
+	var wg sync.WaitGroup
+	for i := 0; i < 10; i++ {
+		// Add them asynchronously to test the
+		// concurrent guarantees of our APIs.
+		wg.Add(1)
+		go func(i int) {
+			defer wg.Done()
```

Comment:
Same Q - what are we actually testing by running concurrent routines here? Perhaps we should Set in one routine and Get in another ?

Reply:
Just one doing a set in one routine and get in another would be too deterministic and non-clashing IMO. However, I like the idea of some Set going on in different goroutines, then Get in the others.

Reply:
Should we do that?

Reply:
Should we do that?

Reply:
@ebuchman now coming back to think of it, doing Gets and Sets in separate goroutines just serves as perhaps a test for race detector, but not actual actionable tests for us because we won't have reliable assertions to make e.g
```go
go func() {
  // Reads here
  // Can't guarantee the number of failed or successful reads
  // as that depends on at least one Set running before majority
  // of the reads here. What do we assert? A percentage of successful
  // reads?
}()

go func() {
   // Writes here
   // Can guarantee that we get a duplicate here
}()
```